### ISSUE TITLE
`hb_buffer_get_direction` may not match the used direction

### ISSUE THREAD

**Issue by kojiishi** [harfbuzz/harfbuzz#3549]
Hi,

I noticed that `hb_buffer_get_direction` may not match the actual used direction, maybe in some cases, tested on Mac.
```sh
% hb-shape /System/Library/Fonts/Geneva.ttf --unicodes=1DDE,A74A --direction=ltr
[uniA74A=1+1861|uni1DDE=0@-371,826+0]
```
`hb_buffer_get_direction` returns `ltr`, but the cluster index is in the reverse order. From the behavior, I'm guessing that the HarfBuzz uses `rtl` to shape for the given Unicode, but `hb_buffer_get_direction` returns the specified direction.

Currently Blink assumes that the buffer is ordered by the cluster index if `hb_buffer_get_direction` is `ltr`, and this mismatch is causing crashes.

The crashes are observed mostly on Mac. They started increasing for the builds between Dec 9th, 2021 to Jan 20th, 2022. Maybe the code has been so before that but such content started appearing, I didn't test past versions.

Is there a way to get the used direction of the `hb_buffer`?

**Comment by kojiishi** [comment]
I think I missed the possibility that the direction change can occur in the middle of a run. So for this test:
```sh
% hb-shape /System/Library/Fonts/Geneva.ttf --unicodes=1DDE,A74A,0041,0042 --direction=ltr
[uniA74A=1+1861|uni1DDE=0@-371,826+0|A=2+1489|B=3+1303]
```
Glyphs are neither in the increasing order nor in the decreasing order, so I guess the question is more complicated.

A large part of the code in Blink are written with the assumption. I'm still thinking how to support this scenario. Is it possible to know if the direction change has occurred in the `hb_buffer`, so that I can add some logic to handle the situation? It'd be great if we can do that without scanning all glyphs.

**Comment by kojiishi** [comment]
Sorry for a noise, I found we're iterating all glyphs before we need to know the direction, so we can detect the situation without adding another loop.

**Comment by kojiishi** [comment]
...and the test case is not about RTL as the two characters in question are Latin. I guess HarfBuzz is smart to reverse the source text if the combining mark is at the beginning, before the base.

**Comment by khaledhosny** [comment]
> ```shell
> % hb-shape /System/Library/Fonts/Geneva.ttf --unicodes=1DDE,A74A,0041,0042 --direction=ltr
> [uniA74A=1+1861|uni1DDE=0@-371,826+0|A=2+1489|B=3+1303]
> ```

This looks like a bug to me, clusters shuld never go like this in level 0 or 1.

**Comment by behdad** [comment]
Might have been fixed by https://github.com/harfbuzz/harfbuzz/commit/4d48fae4f247c789612f6a888aba6abf7906f012

**Comment by behdad** [comment]
Humm. Nope. Broken still.

**Comment by behdad** [comment]
The shaping is broken as well.

**Comment by behdad** [comment]
> Might have been fixed by [4d48fae](https://github.com/harfbuzz/harfbuzz/commit/4d48fae4f247c789612f6a888aba6abf7906f012)

Yeah. Fixed by that. @khaledhosny can you get a release out please?

**Comment by khaledhosny** [comment]
https://github.com/harfbuzz/harfbuzz/releases/tag/4.2.1
